docs/modules/release-notes/pages/0.27.adoc

```diff
@@ -19,34 +21,520 @@ To get started, follow xref:pkl-cli:index.adoc#installation[Installation].#
 
 News you don't want to miss.
 
-=== XXX
+[#typecheck-improvements]
+=== Improved typechecking of `Listing` and `Mapping` types
+
+Typechecking of `Listing` and `Mapping` types has been improved (https://github.com/apple/pkl/pull/628[#628], https://github.com/apple/pkl/pull/725[#725], https://github.com/apple/pkl/pull/740[#740], https://github.com/apple/pkl/pull/752[#752]).
+
+Today, the typecheck `listing: Listing<E>` immediately evaluates all listing elements to check that they have type `E`.
+Likewise, the typecheck `mapping: Mapping<K, V>` immediately evaluates all mapping values to check that they have type `V`.
+
+For example, the typecheck `listing: Listing<Bird>` proceeds as follows:
+
+1. Check that `listing` has type `Listing`
+2. Evaluate each listing element and check that it has type `Bird`
+
+This behavior is different from how the rest of Pkl works.
+Generally, Pkl only evaluates code that affects program output.
+
+For example, consider the following program:
+
+[source,pkl]
+----
+class Bird {
+  name: String
+  canFly: Boolean
+}
+
+local bird: Bird = new {
+  name = "Pidgy"
+  canFly = throw("uh oh")
+}
+
+birdName = bird.name
+----
+
+Even though `bird.canFly` throws an error, the above program succeeds because `bird.canFly` is not part of the program's output and hence is never accessed (Note that `bird` is a _local_ property).
+
+Typechecks of `Mapping` and `Listing` types have been changed to match this behavior.
+Mapping values and listing elements are now only typechecked if and when they are accessed.
+
+This change causes some previously failing programs to evaluate successfully:
+
+[source,pkl]
+----
+local myNumbers: Listing<Int> = new { 1; 2; "uh oh" } 
+result = myNumbers[0]
+----
+In Pkl 0.26 and below, the above program fails with a type mismatch error because element `"uh oh"` is typechecked when `myNumbers` is accessed.
+In Pkl 0.27, the same program succeeds because only element `myNumbers[0]` is part of the program's output, and its typecheck succeeds.
```

Comment:
```suggestion
In Pkl 0.27, the same program succeeds, because only element `myNumbers[0]` is part of the program's output and its typecheck succeeds.
```